theano/gpuarray/subtensor.py

```diff
@@ -1182,3 +1183,43 @@ def infer_shape(self, node, shapes):
             diag_size = T.minimum(dim1, dim2)
         out_shape.append(diag_size)
         return [tuple(out_shape)]
+
+
+class GpuAllocDiag(Subtensor):
+    __props__ = ("offset",)
+
+    def __init__(self, offset=0):
+        self.offset = offset
+
+    def make_node(self, _x):
+        ctx_name = infer_context_name(_x)
+        x = as_gpuarray_variable(_x, ctx_name)
+
+        if x.ndim != 1:
+            raise ValueError('AllocDiag argument must be a vector!', x)
+        
+        return gof.Apply(self, [x], [x.type.__class__(dtype=x.dtype,
+            broadcastable=x.broadcastable)()])
+
+    def perform(self, node, inputs, outputs):
+        (x,) = inputs
+        (z,) = outputs
+
+        dim = x.shape[0] + abs(self.offset)
+        z = gpuarray.zeros((dim, dim), dtype=x.dtype, context=x.context)
+
+        if self.offset <= 0:  # diag in the lower triangle
+            diag_z = z[-self.offset, :(dim + self.offset)]
+        else:  # diag in the upper triangle
+            diag_z = z[:(dim - self.offset), self.offset]
+        diag_z.strides = (sum(z.strides),)
+
+        diag_z[:] = x[:]
+
+    def grad(self, inputs, gout):
+        (input_x,) = inputs
+        return [grad_not_implemented(self, 0, input_x)]
+
+    def infer_shape(self, node, shapes):
+        dim = shapes[0][0] + abs(self.offset)
+        return [dim, dim]
```

Comment:
Returning [[dim, dim]] is also giving an error:

======================================================================
FAIL: test_matrix (theano.gpuarray.tests.test_subtensor.test_gpuallocdiag)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/u/linzhou/Theano/theano/gpuarray/tests/test_subtensor.py", line 207, in test_matrix
    fn = theano.function([x], GpuAllocDiag()(x), mode=mode_with_gpu)
  File "/u/linzhou/Theano/theano/compile/function.py", line 326, in function
    output_keys=output_keys)
  File "/u/linzhou/Theano/theano/compile/pfunc.py", line 486, in pfunc
    output_keys=output_keys)
  File "/u/linzhou/Theano/theano/compile/function_module.py", line 1794, in orig_function
    output_keys=output_keys).create(
  File "/u/linzhou/Theano/theano/compile/function_module.py", line 1474, in __init__
    optimizer_profile = optimizer(fgraph)
  File "/u/linzhou/Theano/theano/gof/opt.py", line 98, in __call__
    return self.optimize(fgraph)
  File "/u/linzhou/Theano/theano/gof/opt.py", line 87, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/u/linzhou/Theano/theano/gof/opt.py", line 235, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/u/linzhou/Theano/theano/gof/opt.py", line 83, in optimize
    self.add_requirements(fgraph)
  File "/u/linzhou/Theano/theano/tensor/opt.py", line 1482, in add_requirements
    fgraph.attach_feature(ShapeFeature())
  File "/u/linzhou/Theano/theano/gof/fg.py", line 541, in attach_feature
    attach(self)
  File "/u/linzhou/Theano/theano/tensor/opt.py", line 1299, in on_attach
    self.on_import(fgraph, node, reason='on_attach')
  File "/u/linzhou/Theano/theano/tensor/opt.py", line 1362, in on_import
    self.set_shape(r, s)
  File "/u/linzhou/Theano/theano/tensor/opt.py", line 1143, in set_shape
    len(s), r.ndim, sio.getvalue()))
AssertionError: Something inferred a shape with 2 dimensions for a variable with 1 dimensions for the variable:
GpuAllocDiag{offset=0} [id A] <GpuArrayType<None>(float32, (False,))> ''   
 |GpuFromHost<None> [id B] <GpuArrayType<None>(float32, (False,))> ''   
   |<TensorType(float32, vector)> [id C] <TensorType(float32, vector)>
----------------------------------------------------------------------